cassiopeia/simulator/Cas9LineageTracingDataSimulator.py

```diff
@@ -0,0 +1,202 @@
+
+    Third, the class accepts a state distribution describing the relative
+    likelihoods of various indels. This is very useful, as it is typical that a
+    handful of mutations are far likelier than the bulk of the possible
+    mutations.
+
+    Finally, the class accepts a silencing rate. This is normally a rare event
+    in which an entire cassette is transcriptionally silenced and therefore
+    not observed.
+
+    The function `overlay_data` will operate on the tree in place and will
+    specifically modify the data stored in the character attributes.
+
+    Args:
+        number_of_cassettes: Number of cassettes (i.e., arrays of target sites)
+        size_of_cassette: Number of editable target sites per cassette
+        mutation_rate: Exponential parameter for the Cas9 cutting rate.
+        state_distribution: Distribution from which to simulate state
+            likelihoods
+        number_of_states: Number of states to simulate
+        silencing_rate: Silencing rate for the cassettes, per node
+    """
+
+    def __init__(
+        self,
+        number_of_cassettes: int = 10,
+        size_of_cassette: int = 3,
+        mutation_rate: float = 0.01,
+        state_distribution: Callable[None, float] = lambda: np.random.exponential(1e-5),
+        number_of_states: int = 100,
+        silencing_rate: float = 1e4,
+    ):
+
+        self.number_of_cut_sites = number_of_cut_sites
+        self.number_of_cassettes = number_of_cassettes
+        self.mutation_rate = mutation_rate
+
+        self.mutation_priors = {}
+        probabilites = [state_distribution() for _ in range(number_of_states)]
+        Z = sum(probabilites)
+        for i in range(number_of_states):
+            self.mutation_priors[i + 1] = probabilites[i] / Z
+
+        self.silencing_rate = silencing_rate
+    
+    def overlay_data(self, tree: CassiopeiaTree):
+        """Overlays Cas9-based lineage tracing data onto the CassiopeiaTree.
+
+        Args:
+            tree: Input CassiopeiaTree
+        """
+        
+        number_of_characters = self.number_of_cassettes * self.number_of_cut_sites
+        for node in tree.depth_first_traverse_nodes(tree.root, postorder=False):
+
+            if tree.is_root(node):
+                tree.set_character_states = [0]*number_of_characters
+                continue
+
+            parent = tree.get_parent(node) 
+            t = tree.get_time(node) - tree.get_time(parent)
+
+            p = 1 - (np.exp(-t * self.mutation_rate))
+    
+            character_array = tree.get_character_states(parent)
+            open_sites = [c for c in range(len(character_array)) if character_array[c] == 0]
+
+            number_of_mutations = np.random.binomial(len(open_sites), p)
+            new_cuts = np.random.choice(open_sites, number_of_mutations)
+            
+            # collapse cuts that are on the same cassette
+            character_array, cuts_remaining = self.collapse_sites(character_array, new_cuts)
```

Comment:
Hmm, this is not how I was expecting a double cut to happen. In my mind, whether a double cut happens or not should not depend on what cuts got _accumulated_ during the lifetime of a cell: a double cut should depend on whether the two cuts happened _close_ together in time.

Reply:
I totally agree - this is not necessarily _exactly_ how double cuts would happen.

However, there are lots of other things that we are not modeling. We're not actually modeling DNA repair, we're not actually modeling how long Cas9 takes to induce a cut, etc. These all undoubtedly play a part in these multiple-site resections. I think a good approximation is to assume that if Cas9 cuts at two sites on the same recorder in a cell's lifetime, we'll introduce a resection event.

Reply:
Sounds good! So long as we document this precisely.